src/crimson/mon/MonClient.cc

```diff
@@ -397,7 +398,10 @@ Connection::handle_auth_done(uint64_t new_global_id,
   secret_t connection_secret;
   int r = auth->handle_response(0, p, &session_key, &connection_secret);
   conn->set_last_keepalive_ack(auth_start);
-  auth_done.set_value(auth_result_t::success);
+  if (auth_done) {
+    auth_done->set_value(auth_result_t::success);
```

Comment:
Is `if (auth_done)` here to prevent `set_value()` to be called again, or it is because `auth_done` is not available in `authenticate_v1()` path?

Reply:
auth_done is std::optional, when auth_done.set_value() is called, followed by auth_done.clear(). then auth_done is not available,  so here if (auth_done) check auth_done is available.  that is my understanding.

Reply:
> Is `if (auth_done)` here to prevent `set_value()` to be called again, or it is because `auth_done` is not available in `authenticate_v1()` path?

latter in general. there is no need to set `auth_done` if nobody is waiting for it, as you pointed out. v2 auth does not use `auth_done` at all. so this change makes this explicit. it's unlikely to set `auth_done` twice in this code path. the set-twice issue is spotted in `Connection::close()`.

Reply:
Thanks, I'm clear now.

Nit, `Connection::handle_auth_done()` can be moved down to "v1 and v2" section

Reply:
@cyx1231st updated and repushed.